diff -p -N --text -r /tmp/patch.inc.get1.S4Lt5u/src/core/nginx.h /tmp/patch.inc.get2.V8jTNf/src/core/nginx.h
*** /tmp/patch.inc.get1.S4Lt5u/src/core/nginx.h	2013-04-29 23:58:29.625942813 +0200
--- /tmp/patch.inc.get2.V8jTNf/src/core/nginx.h	2013-04-29 23:58:32.865958880 +0200
***************
*** 8,15 ****
  #define _NGINX_H_INCLUDED_
  
  
! #define nginx_version         9002
! #define NGINX_VERSION      "0.9.2"
  #define NGINX_VER          "nginx/" NGINX_VERSION
  
  #define NGINX_VAR          "NGINX"
--- 8,15 ----
  #define _NGINX_H_INCLUDED_
  
  
! #define nginx_version         9003
! #define NGINX_VERSION      "0.9.3"
  #define NGINX_VER          "nginx/" NGINX_VERSION
  
  #define NGINX_VAR          "NGINX"
diff -p -N --text -r /tmp/patch.inc.get1.S4Lt5u/src/core/ngx_log.c /tmp/patch.inc.get2.V8jTNf/src/core/ngx_log.c
*** /tmp/patch.inc.get1.S4Lt5u/src/core/ngx_log.c	2013-04-29 23:58:29.625942813 +0200
--- /tmp/patch.inc.get2.V8jTNf/src/core/ngx_log.c	2013-04-29 23:58:32.865958880 +0200
*************** ngx_log_error_core(ngx_uint_t level, ngx
*** 148,156 ****
          return;
      }
  
!     msg -= (7 + err_levels[level].len + 4);
  
!     (void) ngx_sprintf(msg, "nginx: [%V]: ", &err_levels[level]);
  
      (void) ngx_write_console(ngx_stderr, msg, p - msg);
  }
--- 148,156 ----
          return;
      }
  
!     msg -= (7 + err_levels[level].len + 3);
  
!     (void) ngx_sprintf(msg, "nginx: [%V] ", &err_levels[level]);
  
      (void) ngx_write_console(ngx_stderr, msg, p - msg);
  }
diff -p -N --text -r /tmp/patch.inc.get1.S4Lt5u/src/http/ngx_http.c /tmp/patch.inc.get2.V8jTNf/src/http/ngx_http.c
*** /tmp/patch.inc.get1.S4Lt5u/src/http/ngx_http.c	2013-04-29 23:58:29.617942774 +0200
--- /tmp/patch.inc.get2.V8jTNf/src/http/ngx_http.c	2013-04-29 23:58:32.857958837 +0200
*************** ngx_http_server_names(ngx_conf_t *cf, ng
*** 1502,1508 ****
              }
  
              if (rc == NGX_BUSY) {
!             ngx_log_error(NGX_LOG_WARN, cf->log, 0,
                                "conflicting server name \"%V\" on %s, ignored",
                                &name[n].name, addr->opt.addr);
              }
--- 1502,1508 ----
              }
  
              if (rc == NGX_BUSY) {
!                 ngx_log_error(NGX_LOG_WARN, cf->log, 0,
                                "conflicting server name \"%V\" on %s, ignored",
                                &name[n].name, addr->opt.addr);
              }
*************** ngx_http_add_addrs6(ngx_conf_t *cf, ngx_
*** 1870,1877 ****
          if (addr[i].hash.buckets == NULL
              && (addr[i].wc_head == NULL
                  || addr[i].wc_head->hash.buckets == NULL)
!             && (addr[i].wc_head == NULL
!                 || addr[i].wc_head->hash.buckets == NULL))
          {
              continue;
          }
--- 1870,1881 ----
          if (addr[i].hash.buckets == NULL
              && (addr[i].wc_head == NULL
                  || addr[i].wc_head->hash.buckets == NULL)
!             && (addr[i].wc_tail == NULL
!                 || addr[i].wc_tail->hash.buckets == NULL)
! #if (NGX_PCRE)
!             && addr[i].nregex == 0
! #endif
!             )
          {
              continue;
          }
diff -p -N --text -r /tmp/patch.inc.get1.S4Lt5u/src/os/unix/ngx_process_cycle.c /tmp/patch.inc.get2.V8jTNf/src/os/unix/ngx_process_cycle.c
*** /tmp/patch.inc.get1.S4Lt5u/src/os/unix/ngx_process_cycle.c	2013-04-29 23:58:29.625942813 +0200
--- /tmp/patch.inc.get2.V8jTNf/src/os/unix/ngx_process_cycle.c	2013-04-29 23:58:32.869958905 +0200
*************** ngx_single_process_cycle(ngx_cycle_t *cy
*** 291,296 ****
--- 291,301 ----
  {
      ngx_uint_t  i;
  
+     if (ngx_set_environment(cycle, NULL) == NULL) {
+         /* fatal */
+         exit(2);
+     }
+ 
      for (i = 0; ngx_modules[i]; i++) {
          if (ngx_modules[i]->init_process) {
              if (ngx_modules[i]->init_process(cycle) == NGX_ERROR) {
